src/Plugin/Field/FieldWidget/ProductVariationSkuWidget.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace Drupal\commerce\Plugin\Field\FieldWidget;
+
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Field\Plugin\Field\FieldWidget\StringTextfieldWidget;
+
+/**
+ * Plugin implementation of the 'commerce_auto_sku' widget.
+ *
+ * @FieldWidget(
+ *   id = "commerce_auto_sku",
+ *   label = @Translation("Commerce auto SKU"),
+ *   field_types = {
+ *     "string"
+ *   }
+ * )
+ */
+class ProductVariationSkuWidget extends StringTextfieldWidget {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function defaultSettings() {
+    return array(
+      'iniqid_enabled' => TRUE,
+      'more_entropy' => FALSE,
+      'prefix' => 'sku-',
+      'suffix' => '',
+      'size' => 60,
+      'placeholder' => '',
+    ) + parent::defaultSettings();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsForm(array $form, FormStateInterface $form_state) {
+    $none = t('None');
+    $settings = $this->getSettings();
+    $element['iniqid_enabled'] = array(
+      '#type' => 'checkbox',
+      '#title' => t('Enable unique auto SKU values generation'),
+      '#default_value' => $settings['iniqid_enabled'],
+    );
+    $element['more_entropy'] = array(
+      '#type' => 'checkbox',
+      '#title_display' => 'before',
+      '#title' => t('More unique'),
+      '#description' => t('If unchecked the SKU (without prefix and suffix) will look like this: <strong>@short</strong>. If checked, like this: <strong>@long</strong>. <a href=":uniqid_href" target="_blank">Read more</a>', [':uniqid_href' => 'http://php.net/manual/en/function.uniqid.php', '@short' => uniqid(), '@long' => uniqid('', TRUE)]),
+      '#default_value' => $settings['more_entropy'],
+      '#states' => array(
+        'visible' => [':input[name*="iniqid_enabled"]' => ['checked' => TRUE]],
+      ),
+    );
+    $element['prefix'] = array(
+      '#type' => 'textfield',
+      '#title' => t('SKU prefix'),
+      '#default_value' => $settings['prefix'],
+      '#placeholder' => $none,
+    );
+    $element['suffix'] = array(
+      '#type' => 'textfield',
+      '#title' => t('SKU suffix'),
+      '#default_value' => $settings['suffix'],
+      '#placeholder' => $none,
+    );
+    $element['size'] = array(
+      '#type' => 'number',
+      '#title' => t('Size of textfield'),
+      '#default_value' => $settings['size'],
+      '#required' => TRUE,
+      '#min' => 1,
+    );
+    $element['placeholder'] = array(
```

Comment:
I don't see how a placeholder will be useful for this widget. The user should not be entering text should they?